apps/homescreen/js/app.js

```diff
@@ -307,17 +210,134 @@ const SETTINGS_VERSION = 0;
+              return Promise.resolve();
             });
           },
           (e) => {
             console.error('Failed to load first-run script');
-            populateApps();
+            return Promise.resolve();
           }) :
-        populateApps);
+        Promise.resolve()),
+
+      // Populate apps
+      new Promise((resolve, reject) => {
+        var request = navigator.mozApps.mgmt.getAll();
+        request.onsuccess = (e) => {
+          for (var app of request.result) {
+            this.addApp(app);
+          }
+          resolve();
+        };
+        request.onerror = (e) => {
+          console.error('Error calling getAll: ' + request.error.name);
+          resolve();
+        };
+      }),
+
+      // Initialise and populate bookmarks
+      this.bookmarks.init().then(() => {
+        document.addEventListener('bookmarks_store-set', (e) => {
+          var id = e.detail.id;
+          this.bookmarks.get(id).then((bookmark) => {
+            for (var child of this.icons.children) {
+              var icon = child.firstElementChild;
+              if (icon.bookmark && icon.bookmark.id === id) {
+                icon.bookmark = bookmark.data;
+                icon.refresh();
+                return;
+              }
+            }
+            this.addAppIcon(bookmark.data);
+          });
+        });
+
+        document.addEventListener('bookmarks_store-removed', (e) => {
+          var id = e.detail.id;
+          for (var child of this.icons.children) {
+            var icon = child.firstElementChild;
+            if (icon.bookmark && icon.bookmark.id === id) {
+              this.icons.removeChild(child, () => {
+                this.storeAppOrder();
+                this.refreshGridSize();
+                this.snapScrollPosition();
+              });
+              this.metadata.remove(id);
+              return;
+            }
+          }
+        });
+
+        document.addEventListener('bookmarks_store-cleared', () => {
+          for (var child of this.icons.children) {
+            var icon = child.firstElementChild;
+            if (icon.bookmark) {
+              this.icons.removeChild(child);
+            }
+          }
+          this.storeAppOrder();
+          this.refreshGridSize();
+          this.snapScrollPosition();
+        });
+      }, (e) => {
+        console.error('Error initialising bookmarks', e);
+      }).then(() => {
+        return this.bookmarks.getAll().then((bookmarks) => {
+          for (var bookmark of bookmarks) {
+            this.addAppIcon(bookmark.data);
+          }
+        }, (e) => {
+          console.error('Error getting bookmarks', e);
+        });
+      })
+    ]).then(() => {
+      // Add any applications that were missed and are in the startup metadata
+      var id;
+      var pendingIcons = this.pendingIcons;
+      this.pendingIcons = {};
+      for (id in pendingIcons) {
+        this.addAppIcon.apply(this, pendingIcons[id]);
+      }
```

Comment:
Is the code above covered by unit tests? I know the code that removes unknown entries from the DB is, so this one should be unit tested too, if that's not already the case.